kivy/properties.pyx

```diff
@@ -635,18 +635,16 @@ cdef class NumericProperty(Property):
     cdef check(self, EventDispatcher obj, value):
         if Property.check(self, obj, value):
             return True
-        if type(value) not in (int, float, long):
-            raise ValueError('%s.%s accept only int/float/long (got %r)' % (
+        if not isinstance(value, (int, float, long)):
```

Comment:
You should use `isinstance(value, Number)` for number check (https://docs.python.org/3.7/library/numbers.html).

Reply:
Just seeing that `isinstance(value, Number)` would include support for `complex` numbers, but that should be supported because it's a `NumericProperty`?

Reply:
> You should use `isinstance(value, Number)` for number check (https://docs.python.org/3.7/library/numbers.html).

I agree with this one. Thanks!

> Just seeing that `isinstance(value, Number)` would include support for `complex` numbers, but that should be supported because it's a `NumericProperty`?

Exactly, if we also include support for complex numbers via the `Number`
class, it will also accept complex numbers included in a numpy's array.

Reply:
I don't completely agree about using `isinstance(value, Number)`. 

The goal of `NumericProperty` is to accept numbers specifically to be used by widgets as part of their positioning and sizing. It's not to accept arbitrary numbers, because arbitrary numbers cannot be used to set a graphics objects position in opengl. You can see it by the fact that it accepts strings e.g. `'5dp'` and that this kind of issue rarely comes up for users or they find good workarounds. 

So we should just accept numbers that can be used for that kind of purpose. This e.g. precludes complex numbers. But, it is reasonable to accept e.g. numpy floats or ints. If someone does want to use complex numbers for their own purpose, they should just use `ObjectProperty`.

Ultimately, if we accept numpy numbers, they'll have to be converted to normal ints and floats in the convert method, because that's the only kind of number `NumericProperty` users typically know how to use. I'd say it's probably best to keep the current logic of `if type(value) not in (int, float, long)` as a quick path for most numbers and same for strings or lists/tuples (maybe with a try/except for iterables as mentioned below). And then if it all fails try to check explicitly `isinstance(x, int)` and `float` and then convert them to the appropriate type.

Reply:
Test `isinstance(value, Number)` won't break `NumericProperty` for current use cases but expand it. I don't think that @Franccisco or @sanderland want to use numpy floats to assign values for pos/size/color (which are used by graphics) but to have property which check/accept all number values and dispatch event when value gets changed through assign or in-place operations (not supported by `ObjectProperty`). 

Documentation of `NumericProperty` should state that `float` type must be used if property is used to update attributes of graphics instructions, but otherwise all number types are accepted. 

If I remember correctly type `long` is removed from Python3 and `int` should be used in its place. Don't know if the same is valid for Cython.

Reply:
The usecase was something like, there is a widget displaying a numeric value (say, win rate). The .kv handles setting the .text given the NumericProperty value. The value comes from some machine learning model.
It would have been fine to raise a TypeError and tell me to cast it to float (although accepting the numpy float would be nicer, of course). The thing that took quite a bit of time was the error about an 'invalid format'.

Reply:
>Test isinstance(value, Number) won't break NumericProperty for current use cases but expand it.

True, but I don't think we should expand it. `NumericProperty` should be limited to things that can be and should be converted to either a `float` or `int`.

>and dispatch event when value gets changed through assign or in-place operations (not supported by ObjectProperty).

Right, object property doesn't support in-place assign and neither does numeric property?

>Documentation of NumericProperty should state that float type must be used if property is used to update attributes of graphics instructions, but otherwise all number types are accepted.

I very much disagree. Currently, we specifically only accept floats or ints (and longs). Because `NumericProperty` is intended specifically for widgets where we need a float/int for layouts or graphics. If you give it anything else it'll crash currently, and that's how it should be. There's really no reason to have specific string and numeric properties, beyond a basic object property, except for the reason that this is needed for setting widget layouts or similar properties.

We need to keep our API as simple as minimal as possible, until there's demonstarted need to accept more number types. And so far I haven't seen a reason given for expanding numeric property beyond floats/int or things that can be converted to floats/ints.

But the docs should clearly say this that it'll be converted to a float or int.

>long is removed from Python3

Yeah, we should remove it.

Reply:
>The usecase was something like, there is a widget displaying a numeric value (say, win rate). The .kv handles setting the .text given the NumericProperty value. The value comes from some machine learning model.

Why can't you use just a object property? Or maybe convert it to a float before setting it. Or even a string property since you'll be converting it to a string. Either way, there's no benefit in storing it as a numeric property. I think the `NumericProperty` is a tempting distraction in this case, because it's intended for places where you need float/ints as mentioned, and it's mainly useful for internal kivy stuff to convert from e.g. `'5dp'` to a number. For normal stuff I'd highly recommend to use object property instead of string/numeric property.

Perhaps we need to say this in the docs.

Reply:
> I don't completely agree about using isinstance(value, Number).
>
> The goal of NumericProperty is to accept numbers specifically to be used by widgets as part of their positioning and sizing. It's not to accept arbitrary numbers, because arbitrary numbers cannot be used to set a graphics objects position in opengl. You can see it by the fact that it accepts strings e.g. '5dp' and that this kind of issue rarely comes up for users or they find good workarounds.
>
> So we should just accept numbers that can be used for that kind of purpose. This e.g. precludes complex numbers. But, it is reasonable to accept e.g. numpy floats or ints. If someone does want to use complex numbers for their own purpose, they should just use ObjectProperty.

The reason of this Pull Request is because it is more reasonable to use
something like a numpy's float/int/complex in a NumericProperty class,
rather than ObjectProperty. To be honest, I did not know that the
purpose of NumericProperty is related to be used by widgets, and nothing
else.

> Ultimately, if we accept numpy numbers, they'll have to be converted to normal ints and floats in the convert method, because that's the only kind of number NumericProperty users typically know how to use. I'd say it's probably best to keep the current logic of if type(value) not in (int, float, long) as a quick path for most numbers and same for strings or lists/tuples (maybe with a try/except for iterables as mentioned below). And then if it all fails try to check explicitly isinstance(x, int) and float and then convert them to the appropriate type.

I don't agree with converting numpy numbers to normal ints or floats,
because it is an additional task that would go against the purpose of
using numpy numbers (performance and speed).

Reply:
> Test isinstance(value, Number) won't break NumericProperty for current use cases but expand it. I don't think that @Franccisco or @sanderland want to use numpy floats to assign values for pos/size/color (which are used by graphics) but to have property which check/accept all number values and dispatch event when value gets changed through assign or in-place operations (not supported by ObjectProperty).

Exactly. As I said, the idea is to expand NumericProperty. After all,
when I want to use something like a numpy number, the ObjectProperty
class is not the first thing that comes to my mind.

> Documentation of NumericProperty should state that float type must be used if property is used to update attributes of graphics instructions, but otherwise all number types are accepted.

I agree. After all, if you use something that is not a
float/int/list/tuple, you get an error that is not so clear.

> If I remember correctly type long is removed from Python3 and int should be used in its place. Don't know if the same is valid for Cython.

I did not remove it because I thought it was there for a good reason.
As it is deprecated, it should be removed.

Reply:
I checked and both `NumericProperty` and `ObjectProperty` will dispatch events for in-place operations. 

Kivy doesn't completely safeguard graphics attributes when assigning from properties as user can have a `ListProperty` of `numpy.float64` (or values any other type) and try to assign it to:
- pos/size of `Rectangle` instruction
- vertices of `Mesh` instruction
- rgba of `Color` instruction

and those values will get pass to graphics without type check and invalid types will raise errors on graphics side.

Also, element-wise changes are possible like:
`label.color[0] = <instance not float or int>`
as `ColorProperty` nor `ObservableList` do not check for types of each element and Kivy still expects from user to know which types are allowed for graphics. Currently graphics instructions will cast `numpy.float64` to float, so cases like this:
`Window.clearcolor[0] = np.float64(0.5)`
do work. I didn't test but I assume that pos/size of `Rectangle` will also cast to `float` if `numpy.float64` values are assigned. 

Because of its naming and its documentation stating: "Property that represents a numeric value.", user will expect that this property can be used for any number for bind/dispatch feature. Expanding it doesn't break its current usage and it give user full support for numbers which are stored without cast except for strings which are cast to `float`. 

From my view it's up to graphics instructions to state in documentation (some classes already do) which types are compatible/required, like any value which can be cast to `float` is valid and not up to `NumericProperty` to ensure `float/int`.

Reply:
Difference in `NumericProperty` vs `ObjectProperty` is the `convert` method which for `ObjectProperty` returns value unchanged and for expanded `NumericProperty` we would have: 
```
if isinstance(x, Number) or x is None:
    return x
# Next lines would check if `x` is a str first and then collection and return `float`
```
as first two lines in `convert` method. Method `check` is similar as `NumericProperty` checks for `Number` or `str` and `ObjectProperty` checks for `baseclass`, both on top of `Property.check` method.

Since for most common use case, assigning a number value, `NumericProperty` does what it name implies, but `ObjectProperty` is slightly faster. 

Someone can argue that users can use:
`number = ObjectProperty(0, baseclass=numpy.float64)`
or
`number = ObjectProperty(0, baseclass=Number)`
but how many users know about `baseclass` argument? @Franccisco @sanderland

Even with `ObjectProperty` used to store number, user can still pass that number to graphics instruction and if it can be cast to float (or any other type which instruction accepts) it still works as extended `NumericProperty` and makes pointless to accept only `float/int`.

Reply:
I've responded to this in the other thread below, but to summarize the goal of `NumericProperty` is not to accept any number, but to accept float/int or things that can be converted to float/int. I don't think we should expand the API beyond that, except to try to convert to int/float if we can, otherwise raise an error.

>Kivy doesn't completely safeguard graphics attributes when assigning from properties as user can have a ListProperty of numpy.float64 (or values any other type) and try to assign it to

True that not all of the properties guard for this, but still that's unrelated to the design of `NumericProperty` whose goal in kivy is currently float/int to be used for pos etc.

>Even with ObjectProperty used to store number, user can still pass that number to graphics instruction and if it can be cast to float (or any other type which instruction accepts) it still works as extended NumericProperty and makes pointless to accept only float/int

We don't do explicit casting in cython because it is expected only ints or floats will be passed to graphics. It is the user's responsibility to ensure that. In fact we have special code in `blit_buffer` to be able accept numpy arrays because it is not normally accepted. And I don't think forcing conversions to float/int in each graphics instruction is a good idea. So this is the job of `NumericProperty`, or the user if they are using a different property.